lighthouse-cli/bin.ts

```diff
@@ -270,34 +198,80 @@ function handleError(err: LightHouseError) {
-
-      return Promise
-        .race([launchChromeAndRun(urls, config, cliFlags), isSigint])
-        .catch(maybeSigint => {
-          if (maybeSigint === _SIGINT) {
-            return cleanup
-              .doCleanup()
-              .catch(err => {
-                console.error(err);
-                console.error(err.stack);
-              }).then(() => process.exit(_ERROR_EXIT_CODE));
-          }
-          return handleError(maybeSigint);
-        });
+/**
+ * If the requested port is 0, set it to a random, unused port.
+ */
+function initPort(flags: {port: number}): Promise<undefined> {
+  return Promise.resolve().then(() => {
+    if (flags.port !== 0) {
+      log.verbose('Lighthouse CLI', `Using supplied port ${flags.port}`);
+      return;
     }
+
+    log.verbose('Lighthouse CLI', 'Generating random port.');
+    return randomPort.getRandomPort().then(portNumber => {
+      flags.port = portNumber;
+      log.verbose('Lighthouse CLI', `Using generated port ${flags.port}.`);
+    });
   })
 }
 
+/**
+ * Attempts to connect to an instance of Chrome with an open remote-debugging
+ * port. If none is found and the `skipAutolaunch` flag is not true, launches
+ * a debuggable instance.
+ */
+function getDebuggableChrome(flags: {skipAutolaunch: boolean, port: number, selectChrome: boolean}): Promise<ChromeLauncher> {
+  const chromeLauncher = new ChromeLauncher({
+    port: flags.port,
+    autoSelectChrome: !flags.selectChrome,
+  });
+
+  // Kill spawned Chrome process in case of ctrl-C.
+  process.on(_SIGINT, () => {
+    chromeLauncher.kill().then(() => process.exit(_SIGINT_EXIT_CODE), handleError);
+  });
+
+  return chromeLauncher
+    // Check if there is an existing instance of Chrome ready to talk.
+    .isDebuggerReady()
+    .catch(() => {
+      if (flags.skipAutolaunch) {
+        return;
+      }
+
+      // If not, create one.
+      log.log('Lighthouse CLI', 'Launching Chrome...');
+      return chromeLauncher.run();
+    })
+    .then(() => chromeLauncher);
+}
+
+function runLighthouse(url: string,
+                       flags: {port: number, skipAutolaunch: boolean, selectChrome: boolean, output: any, outputPath: string},
+                       config: Object): Promise<undefined> {
+
+  let chromeLauncher: ChromeLauncher;
+  return initPort(flags)
+    .then(() => getDebuggableChrome(flags))
+    .then(chrome => chromeLauncher = chrome)
+    .then(() => lighthouse(url, flags, config))
+    .then((results: Results) => Printer.write(results, flags.output, flags.outputPath))
+    .then((results: Results) => {
+      if (flags.output === Printer.OutputMode[Printer.OutputMode.pretty]) {
+        const filename = `./${assetSaver.getFilenamePrefix({url})}.report.html`;
+        Printer.write(results, 'html', filename);
+      }
+    })
+    .then(() => chromeLauncher.kill())
+    .catch(handleError);
+}
+
+function run() {
+  return runLighthouse(url, cliFlags, config);
+}
+
 export {
-  run,
-  launchChromeAndRun
```

Comment:
seems ok. I consume lighthouse like this in pwmetrics:
https://github.com/paulirish/pwmetrics/blob/master/index.js#L46-L68

i use the ChromeLauncher module and then when debuggerReady resolves I run the lighthouse module.  That all seems OK with these refactors. I tested the PR against it on saturday and all was good.